### PR TITLE
Deprecated TypeMap complex fields

### PR DESCRIPTION
**Pull request by zippolyte**
deprecate:
* `layout` typeMap
* `time` typeMap
* `compute` typeMap
* `sort` typeMap
* `search` typeMap

add:
* `widget_layout` typeList; maxItems: 1
* `live_span` typeString
* `compute_query` typeList; maxItems: 1
* `sort_query` typeList; maxItems: 1
* `search_query` typeString

### CODE REVIEW

File: datadog/resource_datadog_dashboard.go [hunk 1]
```diff
@@ -5152,16 +5303,24 @@ func buildTerraformWidgetEvents(datadogWidgetEvents *[]datadogV1.WidgetEvent) *[
...
+func getWidgetLiveSpanSchema() *schema.Schema {
+	return &schema.Schema{
+		Description:  "The timeframe to use when displaying the widget. One of `10m`, `30m`, `1h`, `4h`, `1d`, `2d`, `1w`.",
+		Type:         schema.TypeString,
+		ValidateFunc: validateEnumValue(datadogV1.NewWidgetLiveSpanFromValue),
+		Optional:     true,
 	}
 }
+
+func getWidgetNestedLiveSpanSchema() *schema.Schema {
```
Comment (therve) [inline datadog/resource_datadog_dashboard.go:5315]: Can you reuse `getWidgetNestedLiveSpanSchema`?

File: datadog/resource_datadog_dashboard.go [hunk 2]
```diff
@@ -5152,16 +5303,24 @@ func buildTerraformWidgetEvents(datadogWidgetEvents *[]datadogV1.WidgetEvent) *[
...
-	return map[string]*schema.Schema{
-		"live_span": {
-			Description:  "The timeframe to use when displaying the widget. One of `10m`, `30m`, `1h`, `4h`, `1d`, `2d`, `1w`.",
-			Type:         schema.TypeString,
-			ValidateFunc: validateEnumValue(datadogV1.NewWidgetLiveSpanFromValue),
-			Optional:     true,
-		},
+func getWidgetLiveSpanSchema() *schema.Schema {
+	return &schema.Schema{
+		Description:  "The timeframe to use when displaying the widget. One of `10m`, `30m`, `1h`, `4h`, `1d`, `2d`, `1w`.",
```
Comment (therve) [inline datadog/resource_datadog_dashboard.go:5308]: The description is outdated, can you add the missing values?

**Review by therve — CHANGES_REQUESTED** [review]
Some minor changes. It looks good but a bit big, I hope we're not missing something.

**Comment by zippolyte** [comment]
I feel confident nothing is missing with the extensive per widget testing

**Review by therve — DISMISSED** [review]

File: datadog/resource_datadog_dashboard.go [hunk 3]
```diff
@@ -1212,7 +1274,12 @@ func buildTerraformAlertGraphDefinition(datadogDefinition datadogV1.AlertGraphWi
 		terraformDefinition["title_align"] = *v
 	}
 	if v, ok := datadogDefinition.GetTimeOk(); ok {
-		terraformDefinition["time"] = buildTerraformWidgetTime(*v)
+		// Set to deprecated field if that's what is used in the config, otherwise, set in the new field
+		if _, ok := k.GetOkWith("time"); ok {
+			terraformDefinition["time"] = buildTerraformWidgetTime(*v)
```
Comment (jbenais) [inline datadog/resource_datadog_dashboard.go:1279]: nit: for consistency and clarity, should we replace the function name by `buildTerraformWidgetTimeDeprecated`

**Review by jbenais — APPROVED** [review]
Looks good to me, tested with old / new fields. Amazing work 🚀

**Review by therve — APPROVED** [review]

**Pull request merged**